packages/bull-vault/src/FlashBull.sol

```diff
@@ -3,38 +3,163 @@ pragma solidity =0.7.6;
  */
-contract FlashBull {
+contract FlashBull is UniBull {
     using StrategyMath for uint256;
 
     BullStrategy private bullStrategy;
 
+    uint256 private constant ONE = 1e18;
+    uint32 private constant TWAP = 420;
+
     /// @dev enum to differentiate between Uniswap swap callback function source
     enum FLASH_SOURCE {
         GENERAL_SWAP,
-        FLASH_DEPOSIT,
+        FLASH_DEPOSIT_CRAB,
+        FLASH_DEPOSIT_COLLATERAL,
         FLASH_WITHDRAW
     }
 
-    // struct FlashDepositData {}
-    // // TODO
+    /// @dev wPowerPerp address
+    address private immutable wPowerPerp;
+    
+    /// @dev weth address
+    address private immutable weth;
+
+    /// @dev usdc address
+    address private immutable usdc;
+
+    /// @dev ETH:wSqueeth Uniswap pool
+    address private immutable ethWSqueethPool;
+
+    /// @dev ETH:USDC Uniswap pool
+    address private immutable ethUSDCPool;
+    
+
+    struct FlashDepositCrabData {
+        uint256 usdcAmount;
+        uint256 crabAmount;
+    }
+
+    struct FlashDepositCollateralData {
+        uint256 crabAmount;
+        uint256 squeethAmount;
+    }
 
     // struct FlashWithdrawData {}
     // // TODO
 
-    constructor(address _bull) {
+    constructor(address _bull, address _powerTokenController) {
         bullStrategy = BullStrategy(_bull);
+        wPowerPerp = IController(bullStrategy.powerTokenController).wPowerPerp();
+        weth = IController(bullStrategy.powerTokenController).weth();
+        usdc = IController(bullStrategy.powerTokenController).quoteCurrency();
+        ethWSqueethPool = IController(bullStrategy.powerTokenController).wPowerPerpPool();
+        ethUSDCPool = IController(bullStrategy.powerTokenController).ethQuoteCurrencyPool();
     }
 
     /**
-     * @notice payable deposit
+     * @notice flash deposit into strategy, providing ETH, selling wSqueeth and dollars, and receiving strategy tokens
+     * @dev this function will execute a flash swap where it receives ETH, deposits, mints, and collateralizes the loan using flash swap proceeds and msg.value, and then repays the flash swap with wSqueeth and USDC
+     * @dev _ethToDeposit must be less than msg.value plus the proceeds from the flash swap
+     * @param _ethToDeposit total ETH that will be deposited in to the strategy which is a combination of msg.value and flash swap proceeds
+     * @param _ethToCrab ETH that will be deposited into the crab component
+     * @param _poolFee Uniswap pool fee
      */
-    function flashDeposit() external payable {}
+    function flashDeposit(uint256 _ethToCrab, uint256 _ethToDeposit, uint24 _poolFee) external payable {
+        uint256 ethUsdPrice = _getTwap(ethUSDCPool, weth, usdc, TWAP, false);
+        uint256 squeethEthPrice = _getTwap(ethWSqueethPool, wPowerPerp, weth, TWAP, false);
+        
+        (uint256 ethInCrab, uint256 squeethInCrab) = bullStrategy.getCrabVaultDetails();
+
+        // TODO: Helper function for Crab-USD price in BullStrategy.sol
```

Comment:
I think we should keep crab USD price separate instead of having this helper function as it will cost more to call